File: src/main/java/com/baidu/algorithm/binarysearch/_74_Search_a_2D_Matrix.java
/**
 * Copyright (C) 2016 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.binarysearch;

/**
 * _74_Search_a_2D_Matrix
 *
 * @author xuhaoran01
 */
public class _74_Search_a_2D_Matrix {

    public boolean searchMatrix(int[][] matrix, int target) {

        if (matrix.length == 0 || matrix[0].length == 0) {
            return false;
        }

        int m = matrix.length, n = matrix[0].length;
        int left = 0, right = m * n - 1;
        while (left <= right) {
            int mid = (left + right) >> 1;
            int row = mid / n;
            int col = mid % n;
            if (matrix[row][col] == target) {
                return true;
            }
            else if (matrix[row][col] < target) {
                left = mid + 1;
            }
            else {
                right = mid - 1;
            }
        }

        return false;
    }
}

File: src/main/java/com/baidu/algorithm/binarysearch/_74_Search_a_2D_Matrix.py
def searchMatrix(matrix, target):
    """
    :type matrix: List[List[int]]
    :type target: int
    :rtype: bool
    """
    if len(matrix) == 0 or len(matrix[0]) == 0:
        return False

    m, n = len(matrix), len(matrix[0])
    left, right = 0, m * n - 1

    while (left <= right):
        mid = (left + right) >> 1
        row, col = mid / n, mid % n

        if matrix[row][col] == target:
            return True
        elif matrix[row][col] < target:
            left = mid + 1
        else:
            right = mid - 1

    return False